### PR TITLE
FEATURE: Group mouse selection

### PR DESCRIPTION
**Pull request by unclecheese**
Adds group selection of items using the mouse (lasso). Hold meta key to concatenate selection.

Resolves: https://github.com/silverstripe/silverstripe-framework/issues/7313

### CODE REVIEW

File: client/src/containers/Gallery/Gallery.js [hunk 1]
```diff
@@ -368,6 +379,14 @@ class Gallery extends Component {
   }
 
   /**
+   * Toggle concatenating selected items based on the key event
+   * @param e
+   */
+  toggleSelectConcat(e) {
+    this.props.actions.gallery.setConcatenateSelect(!!e.metaKey);
```
Comment (flamerohr) [inline client/src/containers/Gallery/Gallery.js]: `e.metaKey` returns a boolean anyway, so we don't need the casting right?

I think you'll need to add a `e.metaKey || e.ctrlKey` check, for mac and windows differences.

File: client/src/state/gallery/GalleryActions.js [hunk 2]
```diff
@@ -32,6 +32,18 @@ export function selectFiles(ids = null) {
 }
 
 /**
+ * Toggles whether group selection is additive o
```
Comment (flamerohr) [inline client/src/state/gallery/GalleryActions.js]: broken sentence in description

File: client/src/containers/Gallery/Gallery.js [hunk 3]
```diff
@@ -820,7 +858,16 @@ class Gallery extends Component {
       sectionConfig: this.props.sectionConfig,
     };
 
-    return <GalleryView {...props} />;
+    return (this.props.view === 'table')
+      ? <TableView {...props} />
+      : <SelectableGroup
```
Comment (flamerohr) [inline client/src/containers/Gallery/Gallery.js]: I think this failed linting, need to wrap in `( )` and indent correctly

File: client/src/state/gallery/GalleryActions.js [hunk 4]
```diff
@@ -32,6 +32,18 @@ export function selectFiles(ids = null) {
...
+ * @param bool
+ * @returns {function(*): *}
```
Comment (flamerohr) [inline client/src/state/gallery/GalleryActions.js]: Docblock:
```
@param {boolean} concat
@returns {function}
```

I don't think it needs the casting for the dispatch call either

File: client/src/containers/Gallery/Gallery.js (hunk 3 above)
Comment (unclecheese) [reply client/src/containers/Gallery/Gallery.js]: My `yarn lint` is coming back empty, and the travis tests have passed. Are you sure about this one?

Comment (flamerohr) [reply client/src/containers/Gallery/Gallery.js]: strange
it is practiced in the majority of the code base too...

for reference, this is what I'm talking about https://github.com/airbnb/javascript/tree/master/react#parentheses

**Comment by flamerohr** [comment]
After testing this out, two quirks that I'm looking to fix:
- `mouseup` only works if done inside the selectable area, kind of weird behaviour when even a pixel outside the selectable area.
- strange padding between the toolbar and gallery area, will look at decreasing the bottom padding on the toolbar and increasing the padding above the gallery area

**Comment by flamerohr** [comment]
Added commit and would be good to merge either https://github.com/unclecheese/react-selectable/pull/38 or https://github.com/unclecheese/react-selectable/pull/33 to fix the mouseup and mousemove quirk

**Comment by flamerohr** [comment]
Good to merge on green :)

**Pull request merged**